File: ruby/binary/binary.rb
class Binary

  def self.to_decimal(bit_string)
    raise ArgumentError.new("String must be valid binary (0's and 1's only)") if !self.is_binary?(bit_string)
    position = 0
    bit_string.reverse.chars.reduce(0) do |acc, bit|
      acc += (1 << position) if bit == "1"
      position += 1
      acc
    end
  end

  def self.is_binary?(bit_string)
    /^[01]+$/ =~ bit_string
  end
end

module BookKeeping
  VERSION = 3
end

File: java/binary/src/main/java/Binary.java
public class Binary {
    private String binary;

    public Binary(String binary) {
        this.binary = binary;
    }

    public int getDecimal() {
        if (!valid()) { return 0; }

        int value = 0;
        int len = binary.length();
        char[] chars = binary.toCharArray();

        int val = 1;
        for( int pos = len-1; pos >= 0; pos-- ) {
            if (chars[pos] == '1') {
                value += val;
            }
            val <<= 1;
        }
        return value;
    }

    private boolean valid() {
        return (binary != null && binary.length() != 0 && binary.matches("^[01]*$"));
    }
}